zero out possible NaN in pytorch.ctc_loss

yqwangustc opened the pull request:
When use ATen version of CTCLoss, we observe indeterminstically having NaN in the output gradient. Initial examination reveal that  1) those NaN happens rarely and only at BLANK position. 2) if we disable large batch handling (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/LossCTC.cu#L552), NaN disappears.
Based on the above observation, I believe this line (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/LossCTC.cu#L560) might incur NaN indeterminstically.
Unfortunately, due to the indeterminstic behavior, I am not able to provide a repo thus pinpoint the exact root cause. Before we find it, we'd like to place a stopgap here to zero out any NaN in gradient.

Note there is a previous PR which zero out infinity only (https://github.com/pytorch/pytorch/pull/16199), unfortunately, NaN is not infinity. so it won't help if there is a NaN in gradient. This PR overloads the zero_infinity option to zero out both infinity and NaN. Considering they are used for the same purpose, I think it is fine to just use "zero_infinity" option at the moment. 

Differential Revision: D15591296

t-vi commented:
I'm skeptical about this without some repro. My intuition is that if it produces invalid NaNs occasonally, it will likely produce wrong non-NaN results. Could you please capture the inputs to CTC loss that produces NaNs (i.e. keep them around and when you get NaN gradients save them)?
If you use 2d targets: is this something that might be helped by #20971 ?

t-vi commented:
Oh, and thanks for investigating this. It's awesome to see that you tracked this down!

yqwangustc commented:
@t-vi yes, creating a consistent repo is the most important thing here. I was able to repo NaN occasionally by injecting gradient check code during training and pdb once there is any NaN in gradient . But unfortunately, supplying the exact same input to CTCLoss again did not produce NaN. So I am not able to give you a consistent repo yet. I am thinking other approach to get a repo. 

Btw, we use 1-d target so #20971 won’t help. Also we notice that once we disable large batch handling, NaN also disappears. 

Since a lot of people’s work pending a numerical stable CTC loss, may I suggest we can merge this PR as a stopgap solution, similar as the purpose of zero_infinity ?

Commit:
zero out possible NaN in pytorch.ctc_loss

Summary:
When use ATen version of CTCLoss, we observe indeterminstically having NaN in the output gradient. Initial examination reveal that  1) those NaN happens rarely and only at BLANK position. 2) if we disable large batch handling (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/LossCTC.cu#L552), NaN disappears.

Based on the above observation, I believe this line (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/LossCTC.cu#L560) might incur NaN indeterminstically.

Unfortunately, due to the indeterminstic behavior, I am not able to provide a repo thus pinpoint the exact root cause. Before we find it, we'd like to place a stopgap here to zero out any NaN in gradient.

Differential Revision: D15591296

fbshipit-source-id: 6865e8da58df69a740f649635c7cc49253b24621

t-vi commented:
OK, so it's something different. I would indeed want to get us to the situation of `zero_infinity` - where we clearly understand why we're getting NaNs and so can target avoiding them.
I would expect that there would be something about the inputs when the grads turn NaN, but I'd really like to know what it is.

t-vi commented:
The other thing I'm wondering about: Do you get the NaN in the forward or the backward? Because you also delete them in the forward, even though you describe observing them in the gradient only.

t-vi commented on aten/src/ATen/native/LossCTC.cpp:
```diff
@@ -210,7 +210,8 @@ Tensor ctc_loss_backward_cpu_template(const Tensor& grad_out, const Tensor& log_
       scalar_t nll = neg_log_likelihood.accessor<scalar_t, 1>()[b];
-      if (zero_infinity &&  nll == std::numeric_limits<scalar_t>::infinity()) {
+      if (zero_infinity &&
+          (nll == std::numeric_limits<scalar_t>::infinity() || std::isnan(nll))) {
```
You mention the gradient becoming NaN, yet here you add a condition for `nll`. Have you actually observed `nll` becoming NaN? What were the inputs when that happened? Is this deterministic?

yqwangustc replied:
No, i never observe NaN in CPU kernel. This part of code is just to make `zero_infinity` behaves in the same manner as the GPU part. And I am OK to remove it.

t-vi commented on aten/src/ATen/native/LossCTC.cpp:
```diff
@@ -364,6 +365,7 @@ Tensor ctc_loss(const Tensor& log_probs, const Tensor& targets, IntArrayRef inpu
     res = std::get<0>(at::_ctc_loss(log_probs, targets, input_lengths, target_lengths, BLANK, zero_infinity));
     if (zero_infinity) {
       res = at::where(res == Scalar(std::numeric_limits<double>::infinity()), at::zeros({}, res.options()), res);
+      res = at::where(res != res, at::zeros({}, res.options()), res);
```
Similar to the above, this is on the output of the forward, so I'm not sure it matches your observation.

yqwangustc replied:
similar comments as above.

t-vi replied:
But this is about forward vs. backward, not about CPU vs. GPU.

t-vi replied:
Per your comment below, I think this would not have any effect (until the inputs are NaN), so I think we don't need it.

t-vi commented on aten/src/ATen/native/cuda/LossCTC.cu:
```diff
@@ -571,6 +571,8 @@ Tensor ctc_loss_backward_gpu_template(const Tensor& grad_out, const Tensor& log_
     if (zero_infinity) {
       grad = at::where(neg_log_likelihood.view({1, batch_size, 1}) == Scalar(INFINITY), at::zeros({}, grad.options()), grad);
+      grad = at::where(grad != grad, at::zeros({}, grad.options()), grad);
+      // to filter out NaN value
```
Now, I would like to see where this is actually happening. You describe (and I can believe that) that it's not necessarily reproducible, but as mentioned above, I think we need to understand what are the causes and I would half expect that the result will be wrong in the cases when otherwise NaN-producing inputs produce non-NaN results.
Can you capture a sample of inputs for which the gradient turns NaN, please?

yqwangustc replied:
@t-vi , yes this is the main fix I want to perform -- GPU kernels on backward, all those CPU code change is to make the `zero_infinity` option behaves consistently. As I said I can occasionally capture NaN, however, it is not reproducible consistently (the input I captured results in NaN, but next time I supply them to the CTCLoss, NaN is gone) and I am working on it to produce a true example (trying different approaches now ...) 

I understand we need to find the root cause of occasional NaN, and I fully agree with you that we need to find out the root cause to fix it permanently.  However, since this NaN issue is blocking a lot of other's work (including me, my colleagues and a lot of other people in PyTorch form), may I suggest that we can patch this for the time being ? I will definitely work with you to find out the real solution afterwards.

t-vi replied:
Could you check whether the gradient is correct when it isn't NaN?

Sorry, but without an example, I'm not enthused about zeroing NaNs. Note that most of the people on the forums ended up having errors in their inputs (non-log-probability inputs, invalid targets...). After seeing so many of those I'll have to see an example where the code goes wrong to add bandaids.

t-vi reviewed (dismissed):
Along the lines of the comments above, I think we need some form of repro / example that causes this, and ideally a test.
I have a strong preference for a more proper - i.e. finding and fixing the apparent algorithmic defect - fix.

yqwangustc commented:
@t-vi 

> The other thing I'm wondering about: Do you get the NaN in the forward or the backward? Because you also delete them in the forward, even though you describe observing them in the gradient only.

sorry that I miss this question. I observe NaN appears in gradient first( no NaN in the input) and loss is a normal number in that case. After gradient has NaN, it propagates to model parameters, and then input (to CTCLoss) in the next mini-batch and then everywhere.

ezyang commented:
@t-vi @yqwangustc and I discussed this offline and decided that we should land this patch to stem the bleeding (but we should continue to investigate what the real problem is.)

ezyang commented:
I have one more request, which is that `zero_infinities` documentation be updated to say that it also zeros NaNs, as a quirk of the current implementation (link to the bug or PR). (I know eventually we will probably want to change it back, but if we forget and this ends up in the release, I'd like the docs to be accurate.)

ezyang commented:
Related: #14335 (no code was provided in the reports but it looks similar)

yqwangustc commented:
A bit more information on the input causing NaN. 

- lprobs.shape
torch.Size([984, 16, 42])
- targets_flat.shape
torch.Size([1316])
- input_length 
tensor([984, 956, 943, 924, 921, 904, 901, 897, 864, 824, 717, 710, 695, 666, 578, 552], device='cuda:0')
- target_lengths
 tensor([ 96,  78,  85, 140, 107,  88, 126,  95,  70,  52,  86,  38,  88,  69, 60,  38], device='cuda:0')
- blank_id = 41

and the loss is a norm floating point, so forward pass is fine. grad[983, 1:10, 41] has all the NaN value, i.e., all the NaN appears at BLANK posterior but they are just padded position in time sequence. Switching to small batch size condition does not yield any NaN, so I am checking grad before/after this line https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/LossCTC.cu#L560 

 @t-vi  will these information help you to identify something ?

t-vi commented:
Thank you for the analysis! This is extremely useful.
So it occurs in the padded part. In this `log_alpha`/`log_beta` will be `-infinity`. I don't think the `as_strided` is particularly problematic, but we should really replace it with `.slice`.
Now if `log_probs` is infinite or NaN (which should not have an effect in the padded part), it'll give NaN (either from NaN or from `inf - inf`).

Thanks again for tracking this down!

So I'm happy with your fix, maybe we can add a comment that this is in the padding and that when the blank calculation is moved to a kernel, properly ignoring those bits will be better.
I'd have a preference to not change the forward (the two other bits), but only zero out the NaNs in the backward.

yqwangustc commented:
@t-vi  

Actually, after many iterations,  I think I found the real root cause now ! 

I start to notice after `log_beta = at::empty_like(log_alpha);` there is already some `nan` in log_beta. Initially I though this might be fine, since all the (b, t, s) cases will be covered in the GPU kernel `ctc_loss_backward_log_beta_gpu_kernel`. However, I found two exceptions: 

- if `s=2*target_length + 1`, since the condition in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/LossCTC.cu#L342 is `s > 2*target_length+1` not `s>=2*target_length+1` (really tricky here !), any element in `log_beta[b, :, 2*target_lengths[b] + 1]`  possibly not initialized. 

- For `t=max_input_length - 1`, I didn't see anywhere in the code to assign value, so `log_beta[b, max_input_length - 1, :]` is also possibly not initialized. 

These are consistent with the pattern of either NaN or some widely large value (e.g., -1.03845089e+34)  I saw in `log_beta`. 

So suggest change: 

- at https://github.com/pytorch/pytorch/blame/2ee2d78a29a583e3f40cd27828f02f6ef627a01c/aten/src/ATen/native/cuda/LossCTC.cu#L520, I am not sure why we use `empty_like`, changing it to `full_like(log_alpha, neginf)` solve all the problems! 


@t-vi  Please kindly correct me if I am wrong. I am running more experiments, but I am quite confident this will permanent fix the problem, :-)

t-vi commented:
He, shaking that function really has the bugs falling out... Awesome work!
I'd probably stick with empty (saves calling a kernel to initialize the memory) if we can make sure we write all of them.
- So in your first observation `>` should be `>=` clearly. 
- The `t = max_input_length - 1` case should be handled by extending the s-loop. I was thinking that `t = input_length - 1` was already covered and did not consider that we would end up using the uninitialized value.
  Then `t >= input_length` would cover that (unless `input_length == max_input_length`).

But I think we might still have to zero "out of sequence" gradients due to the use of log_probs.

t-vi commented:
Do you want help to update the PR? I'd be glad to help code something up, but you did all the hard work already...

yqwangustc commented:
@t-vi  new PR is on the way and will send it out soon. sorry that just wake up, :-)

t-vi commented:
Supercool! Thanks!

yqwangustc commented:
Abandon this PR in favor of pytorch/pytorch #21392
